cpp/merkletree/sparse_merkle_tree.h

```diff
@@ -0,0 +1,243 @@
+//      p:"00"            |
+//      v:"aa"      0_____|_____1
+//                  |           |
+//                p:"10"      p:"11"
+//                v:"hi"      v:"to"
+//
+// * Calculating the root hash
+// Calculating the root of the tree is similar to a regular MerkleTree, but is
+// optimised by cribbing the value of "missing" nodes from a simple cache. This
+// removes the need to calculate the vast majority of nodes from scratch.
+//
+// TODO(alcutter): LOTS!
+//
+// This class is thread-compatible, but not thread-safe.
+class SparseMerkleTree {
+ public:
+  // The constructor takes a pointer to some concrete hash function
+  // instantiation of the SerialHasher abstract class.
+  // Takes ownership of the hasher.
+  explicit SparseMerkleTree(SerialHasher* hasher);
+
+  // Length of a node (i.e., a hash), in bytes.
+  virtual size_t NodeSize() const {
+    return treehasher_.DigestSize();
+  };
+
+  // Number of non-default leaves in the tree.
+  virtual size_t LeafCount() const {
+    LOG(FATAL) << "Not implemented.";
+  }
+
+  // The |leaf|th leaf hash in the tree. Indexing starts from 1.
+  std::string LeafHash(size_t leaf) const {
+    if (leaf == 0 || leaf > LeafCount())
+      return std::string();
+    return Node(0, leaf - 1);
+  }
+
+  // Return the leaf hash, but do not append the data to the tree.
+  virtual std::string LeafHash(const std::string& data) const {
+    return treehasher_.HashLeaf(data);
+  }
+
+  // Add a new leaf to the hash tree. Stores the hash of the leaf data in the
+  // tree structure at path |Hash(key)|, does not store the data itself.
+  //
+  // @param data Binary input blob
+  // @param path Binary path of node to set.
+  virtual void SetLeafByHash(const std::string& key, const std::string& data);
+
+  // Add a new leaf to the hash tree. Stores the hash of the leaf data in the
+  // tree structure, does not store the data itself.
+  //
+  // @param data Binary input blob
+  // @param path Binary path of node to set.
+  virtual void SetLeaf(const Path& path, const std::string& data);
+
+  // Get the current root of the tree.
+  // Update the root to reflect the current shape of the tree,
+  // and return the tree digest.
+  //
+  // Returns the hash of an empty string if the tree has no leaves
+  // (and hence, no root).
+  virtual std::string CurrentRoot();
+
+  // Get the Merkle path from leaf to root.
+  //
+  // Returns a vector of node hashes, ordered by levels from leaf to root.
+  // The first element is the sibling of the leaf hash, and the last element
+  // is one below the root.
+  // Returns an empty vector if the tree is not large enough
+  // or the leaf index is 0.
+  //
+  // @param leaf the index of the leaf the path is for.
+  std::vector<std::string> PathToCurrentRoot(size_t leaf);
+
+  std::string Dump() const;
+
+ private:
+  // WARNING WARNING WARNING
+  // 64 < 256 !
+  // WARNING WARNING WARNING
+  // TODO(alcutter): BIGNUM probably.
+  typedef uint64_t IndexType;
```

Comment:
We did mention an entirely different implementation... While I don't advise doing it, I think it would be handy to have a think about "what if this was storing its data in LevelDB?". And the other way sounds like it would fit that too, and still behave fairly reasonably in-memory?

As a reminder, the idea was to store things in an `std::map`, with the full path as the key. That could use more data, but since the entries would be ordered, I think it would be possible to lazily compute the internal nodes (possibly every time)?